c++/src/kj/async-inl.h

```diff
@@ -306,18 +314,97 @@ class PromiseNode: private AsyncObject {
   private:
     Event* event = nullptr;
   };
+
+private:
+  PromiseArena* arena = nullptr;
+  // If non-null, then this PromiseNode is the last node allocated within the given arena, and
+  // therefore owns the arena. After this node is destroyed, the arena should be deleted.
+  //
+  // PromiseNodes are allocated within the arena starting from the end, and `PromiseNode`s
+  // allocated this way are required to have `PromiseNode` itself as their leftmost inherited type,
+  // so that the pointers match. Thus, the space in `arena` from its start to the location of the
+  // `PromiseNode` is known to be available for subsequent allocations (which should then take
+  // ownership of the arena).
+
+  friend class PromiseDisposer;
 };
 
 class PromiseDisposer {
 public:
   static void dispose(PromiseNode* node) {
+    PromiseArena* arena = node->arena;
     node->destroy();
+    delete arena;
+  }
+
+  template <typename T, typename... Params>
+  static kj::Own<T, PromiseDisposer> alloc(Params&&... params) {
+    // Implements allocPromise().
+    T* ptr;
+    if (sizeof(T) > sizeof(PromiseArena)) {
+      // Node too big, fall back to regular heap allocation.
+      ptr = new T(kj::fwd<Params>(params)...);
+    } else {
+      // Start a new arena.
+      auto* arena = new PromiseArena;
+      ptr = reinterpret_cast<T*>(arena + 1) - 1;
+      ctor(*ptr, kj::fwd<Params>(params)...);
```

Comment:
Not sure if this is a concern, but if ctor throws, would this leak the arena?

Reply:
Hmm actually this does seem to be a problem. Although none of the constructors called by `append()` will throw, there are some constructors called by `alloc()` than can throw (rejecting invalid arguments). In fact, the reason I have not marked `alloc()` as `noexcept` is because it broke a test that was explicitly checking for such an exeption.

I guess I should probably mark `alloc()` as `noexcept` as well, and accept that the exception in question becomes fatal. I think that's probably better than the code bloat needed to conditionally delete the arena on exceptions.

Reply:
I suppose another possibility is just to have an arena always preallocated, and only claim and reallocate it if there is no exception.  But I imagine that has its own set of pitfalls.